### PR TITLE
Use sample_weight in the metric calculations in model validations

### PR DESCRIPTION
**Pull request by ryan-deak-zefr**
# Purpose

To correct metric calculations in model validation (and cross validation) by attempting to use `sample_weight` in the metric calculations when `sample_weight` are supplied in `fit_params`.

#### Reference Issues/PRs

Fixes #4632.  See also #10806.

#### What does this implement/fix? Explain your changes.

This change always uses the same `sample_weight` used in training to weight the metric calculations in `_fit_and_score` and functions called by `_fit_and_score`, so it should be thought of changing the default behavior of things like cross validation.  This is unlike PR #10806 in that it is isolated to `_validation.py` and doesn't touch `_search.py`.  This is very intentional.

#### Any other comments?

This PR provides a simple test showing how unweighted metrics have very unfavorable consequences when models are trained with importance weights but not scored with the same weights.  It is shown in the test that the current omission of `sample_weight` in metrics calculations considerably overestimates the accuracy (0.5 accuracy vs  9.99999 * 10^-7).

### CODE REVIEW

File: sklearn/model_selection/_validation.py [hunk 1]
```diff
@@ -637,6 +650,65 @@ def _multimetric_score(estimator, X_test, y_test, scorers):
...
+    sample_weight : array-like, shape (y)
+        sample weights to use during metric calculation.  May be None.
+
+    Returns
+    -------
+    score : float
+        Score returned by ``scorer`` applied to ``X`` and ``y`` given
+        ``sample_weight``.
+    """
+    if sample_weight is None or np.all(sample_weight == 1):
```
Comment (ryan-deak-zefr) [inline sklearn/model_selection/_validation.py:694]: This `np.all(sample_weight == 1)` check is here to make it so that if `sample_weight` are all ones, then it is the same as omitting.  This allows us to use scorers that don't support `sample_weight`.

File: sklearn/model_selection/_validation.py [hunk 2]
```diff
@@ -637,6 +650,65 @@ def _multimetric_score(estimator, X_test, y_test, scorers):
...
+        try:
+            if y is None:
+                score = scorer(estimator, X, sample_weight=sample_weight)
+            else:
+                score = scorer(estimator, X, y, sample_weight=sample_weight)
+        except TypeError as e:
+            if 'sample_weight' in str(e):
+                raise TypeError(
+                    (
+                        "Attempted to use 'sample_weight' for training "
```
Comment (ryan-deak-zefr) [inline sklearn/model_selection/_validation.py:714]: Repackages error with informative error message.

**Review by zexuan-zhou — APPROVED** [review]
Love the error handling part that deal with some scorers not supporting `sample_weights`

File: sklearn/model_selection/_validation.py [hunk 3]
```diff
@@ -637,6 +654,65 @@ def _multimetric_score(estimator, X_test, y_test, scorers):
...
+        Score returned by ``scorer`` applied to ``X`` and ``y`` given
+        ``sample_weight``.
+    """
+    if sample_weight is None or np.all(sample_weight == 1):
+        if y is None:
+            score = scorer(estimator, X)
+        else:
+            score = scorer(estimator, X, y)
+    else:
+        try:
```
Comment (ryan-deak-zefr) [inline sklearn/model_selection/_validation.py:700]: Is there any benefit to passing `sample_weight` as a keyword argument?  E.g.

```python
kwargs = { 'sample_weight': sample_weight}

# ...

score = score(estimator, X, y, **kwargs)
```

The reason I ask is that when I found the other PR that did importance weighting (https://github.com/scikit-learn/scikit-learn/pull/10806/files#diff-60033c11a662f460e1567effd5faa6f0R584) they had this.  I don't know if that's important or not.  

What does everyone think?

File: sklearn/model_selection/tests/test_validation.py [hunk 4]
```diff
@@ -1582,3 +1582,68 @@ def two_params_scorer(estimator, X_test):
...
+        return_estimator=False,
+        error_score='raise-deprecating'
+    )
+
+    train_metric = res[0][metric_name]
+    test_metric = res[1][metric_name]
+
+    # Unnecessary extra test to illustrate that this is not the desired
+    # metric value (but previously what has been historically returned).
+    assert train_metric != np.sum(y[train]) / y[train].shape[0]
```
Comment (ryan-deak-zefr) [inline sklearn/model_selection/tests/test_validation.py:1640]: This is unnecessary, but I like it for illustrative purposes.

**Comment by vecchp** [comment]
This looks good.

**Review by vecchp — APPROVED** [review]
+1
